pkg/istio/mesh_config.go

```diff
@@ -0,0 +1,223 @@
+package istio
+
+import (
+	"fmt"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+	istiomeshv1alpha1 "istio.io/api/mesh/v1alpha1"
+	istioapiv1alpha1 "istio.io/api/operator/v1alpha1"
+	iopv1alpha1 "istio.io/istio/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/istio/pkg/util/protomarshal"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ExtAuthorizerName = "kuadrant-authorization"
+)
+
+// ConfigWrapper wraps the IstioOperator CRD or ConfigMap
+type ConfigWrapper interface {
+	GetConfigObject() client.Object
+	GetConfig() (*istiomeshv1alpha1.MeshConfig, error)
+	UpdateConfig(updateFunc func(meshConfig *istiomeshv1alpha1.MeshConfig) bool) (bool, error)
```

Comment:
This function seems to to have been thought to receive either `RemoveKuadrantAuthorizerFromConfig` or the return value of `AddKuadrantAuthorizerToConfig(string)`. I was thinking... These are the kind of abstraction that this interface can provide.

E.g.

The following are high-level abstractions that don't have to belong to the `istio` package IMO:

```go
type KuadrantAuthorizer interface {
  GetNamespace() string
}

type ConfigWrapper interface {
  HasKuadrantAuthorizer(KuadrantAuthorizer) bool
  RegisterKuadrantAuthorizer(KuadrantAuthorizer) error
  UnregisterKuadrantAuthorizer(KuadrantAuthorizer) error
  GetConfigObject() client.Object
}
```

Then the `istio` package implements the interface – for the `IstioOperator`, `IstioConfigMap` and `OpenShiftServiceMeshControlPlane` specific types. The following become all details that the KuadrantReconciler doesn't need to know about:
- `MeshConfigToStruct(config *istiomeshv1alpha1.MeshConfig) (*structpb.Struct, error)`
- `MeshConfigFromString(config string) (*istiomeshv1alpha1.MeshConfig, error)`
- `ExtensionProvidersFromMeshConfig(config *istiomeshv1alpha1.MeshConfig) (extensionProviders []*istiomeshv1alpha1.MeshConfig_ExtensionProvider)`
- `CreateKuadrantAuthorizer(namespace string) *istiomeshv1alpha1.MeshConfig_ExtensionProvider`

Maybe even:
- `HasKuadrantAuthorizer(extensionProviders []*istiomeshv1alpha1.MeshConfig_ExtensionProvider) bool`

WDYT?

Reply:
I agree! from the beginning was a bit troubled by the scope of each. Thanks!